src/librender/scene.cpp

```diff
@@ -172,43 +172,117 @@ Scene<Float, Spectrum>::ray_test(const Ray3f &ray, Mask active) const {
+#else
+    UInt32 index = ek::min(UInt32(index_sample * (ScalarFloat) emitters_size),
+                           (uint32_t) emitters_size - 1);
+    // pdf = 1 / emitters_size  =>  sampling_weight = emitters_size
+    return { index, emitters_size };
+#endif
+}
+
+MTS_VARIANT std::tuple<typename Scene<Float, Spectrum>::UInt32, Float, Float>
+Scene<Float, Spectrum>::sample_emitter_reuse(Float index_sample,
+                                             Mask active) const {
+    MTS_MASKED_FUNCTION(ProfilerPhase::SampleEmitter, active);
+
+    size_t emitters_size = m_emitters.size();
+    // Fast paths if there is less than two emitters
+    if (unlikely(emitters_size == 0))
+        return { UInt32(-1), 0.f, index_sample };
+    if (emitters_size == 1)
+        return { 0, 1, index_sample };
+
+#if 0
+    // Randomly pick an emitter
+    auto [index, emitter_pdf, sample_re] = m_emitter_distr.sample_reuse_pmf(index_sample);
+    return { index, ek::rcp(emitter_pdf), sample_re };
+#else
+    UInt32 index = ek::min(UInt32(index_sample * (ScalarFloat) emitters_size),
+                           (uint32_t) emitters_size - 1);
+    // Rescale sample to lie in [0,1) again
+    index_sample = (index_sample - index / (ScalarFloat) emitters_size) * emitters_size;
+    // pdf = 1 / emitters_size  =>  sampling_weight = emitters_size
+    return { index, emitters_size, index_sample };
+#endif
+}
+
 MTS_VARIANT std::pair<typename Scene<Float, Spectrum>::DirectionSample3f, Spectrum>
 Scene<Float, Spectrum>::sample_emitter_direction(const Interaction3f &ref, const Point2f &sample_,
                                                  bool test_visibility, Mask active) const {
     MTS_MASKED_FUNCTION(ProfilerPhase::SampleEmitterDirection, active);
-
     using EmitterPtr = ek::replace_scalar_t<Float, Emitter*>;
 
     Point2f sample(sample_);
     DirectionSample3f ds;
     Spectrum spec;
 
     size_t emitters_size = m_emitters.size();
-
     if (likely(emitters_size > 0)) {
         if (emitters_size == 1) {
             // Fast path if there is only one emitter
             std::tie(ds, spec) = m_emitters[0]->sample_direction(ref, sample, active);
             ek::schedule(ds, spec);
         } else {
-            ScalarFloat emitter_pdf = 1.f / emitters_size;
-
             // Randomly pick an emitter
-            UInt32 index =
-                ek::min(UInt32(sample.x() * (ScalarFloat) emitters_size),
-                    (uint32_t) emitters_size - 1);
-
-            // Rescale sample.x() to lie in [0,1) again
-            sample.x() = (sample.x() - index*emitter_pdf) * emitters_size;
-
-            EmitterPtr emitter = ek::gather<EmitterPtr>(m_emitters_ptr.data(), index, active);
+            auto [index, emitter_weight, sample_re] =
+                sample_emitter_reuse(sample.x(), active);
+            sample.x() = sample_re;
+            // Account for the discrete probability of sampling this emitter
+            ds.pdf *= ek::rcp(emitter_weight);
```

Comment:
This line seems weird just from a convention point of view (Monte Carlo weights can contain all sorts of things that are not necessarily PDFs). Can we instead return the PDF and then multiply by it here? And then `rcp()` that for the weight.

Reply:
I agree that this line looks wrong because it silently assumes that `sample weight == 1 / pdf`. That being said, I would prefer keeping `sample_emitter`'s return value a sampling weight because (as far as I know), every other method starting with `sample_` returns a sampling weight.
Instead, how about: `ds.pdf = pdf_emitter(index);`?

Reply:
Also, I moved back these two lines (update of `ds.pdf` and `spec`) below the `std::tie()`, otherwise their effect was immediately ignored.

Reply:
Your suggestion sounds good.